File: 2017/day16/python/part2.py
def check_cycles():
    input = open('../input.txt', 'r').readline()
    moves = input.split(',')
    programs = ['a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j', 'k', 'l', 'm', 'n', 'o', 'p']
    idx = 1
    while True:
        programs = part1(programs, moves)

        if ("".join(programs) == 'abcdefghijklmnop'):
            print(idx)
            break

        idx += 1
        
def part2():
    input = open('../input.txt', 'r').readline()
    moves = input.split(',')
    programs = ['a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j', 'k', 'l', 'm', 'n', 'o', 'p']
    modulo = 1_000_000_000 % 42 # For our inputs, the dance cycles every 42 iterations
    for i in range(modulo):
        programs = part1(programs, moves)

        if i == 0:
            print("".join(programs))

    print("".join(programs))

def part1(programs, moves):
    for move in moves:
        move_type = move[0]
        if move_type == 's':
            spin_num = int(move[1:])
            end_part = programs[-spin_num:]
            beginning_part = programs[:-spin_num]
            programs = end_part + beginning_part
        elif move_type == 'x':
            coordinates = move[1:].split('/')
            position_1 = int(coordinates[0])
            position_2 = int(coordinates[1])
            programs[position_1], programs[position_2] = programs[position_2], programs[position_1]
        elif move_type == 'p':
            coordinates = move[1:].split('/')
            char_1 = coordinates[0]
            position_1 = programs.index(char_1)
            char_2 = coordinates[1]
            position_2 = programs.index(char_2)
            programs[position_1], programs[position_2] = programs[position_2], programs[position_1]

    return programs

part2()

File: 2017/day16/rust/part2.rs
use std::fs;

fn main() {
    part2();
}

fn part2() {
    let input = fs::read_to_string("input.txt")
        .expect("Something went wrong reading the file");

    let moves: Vec<&str> = input.trim().split(',').collect();
    let mut programs: [char; 16] = ['a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j', 'k', 'l', 'm', 'n', 'o', 'p'];
    let initial_state = "abcdefghijklmnop";
    let mut cycle_length = 1;

    // Look for cycle in program positions
    loop {
        part1(&mut programs, &moves);
        if programs.iter().collect::<String>() == initial_state {
            break;
        }
        cycle_length += 1;
    }

    // Now we don't need to run the dance a billion times, just 1,000,000,000mod(cycle_length) times
    for i in 0..(1_000_000_000 % cycle_length) {
        part1(&mut programs, &moves);

        if i == 0 {
            println!("{}", programs.iter().collect::<String>());
        }
    }
    
    println!("{}", programs.iter().collect::<String>());
}

fn part1(programs: &mut [char; 16], moves: &Vec<&str>) {
    for &move_str in moves {
        let move_type = move_str.chars().next().unwrap();
        let rest_of_move = &move_str[1..];

        match move_type {
            's' => {
                let spin_num: usize = rest_of_move.parse().expect("Invalid spin number");
                programs.rotate_right(spin_num);
            }
            'x' => {
                let mut coordinates = rest_of_move.split('/');
                let pos1: usize = coordinates.next().unwrap().parse().expect("Invalid position");
                let pos2: usize = coordinates.next().unwrap().parse().expect("Invalid position");
                
                programs.swap(pos1, pos2);
            }
            'p' => {
                let mut coordinates = rest_of_move.split('/');
                let char1 = coordinates.next().unwrap().chars().next().unwrap();
                let char2 = coordinates.next().unwrap().chars().next().unwrap();
                let pos1 = programs.iter().position(|&c| c == char1).expect("Character not found");
                let pos2 = programs.iter().position(|&c| c == char2).expect("Character not found");

                programs.swap(pos1, pos2);
            }
            _ => {
                // Handle unknown move types
                eprintln!("Unknown move type: {}", move_type);
            }
        }
    }
}